crates/holochain_types/src/dht_op.rs

```diff
@@ -929,8 +929,13 @@ pub struct OpOrder {
 
 impl std::fmt::Display for OpOrder {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        let (secs, nsecs) = self.timestamp.as_seconds_and_nanos();
-        write!(f, "{}{:019}{:010}", self.order as u8, secs, nsecs)
+        write!(
+            f,
+            "{}{:019}",
+            self.order as u8,
+            // clamp unrealistic negative timestamps to 0
+            i64::max(0, self.timestamp.as_micros())
```

Comment:
why clamp? if someone's generating bad data and injecting it, don't we want to see the truth of what the ops say, rather than arbitrarily hiding it?

Reply:
@neonphog it's because we depend on lexical ordering, and negative numbers are not compatible with lexical ordering, if nothing else due to the minus sign

Reply:
I'm pretty sure the default implementation of `PartialOrd, Ord` on both [OpOrder](https://github.com/holochain/holochain/blob/2e2675bb82918c6c3d25e60cb925ea3a58a9e0be/crates/holochain_types/src/dht_op.rs#L924) delegating to [Timestamp](https://github.com/holochain/holochain/blob/develop/crates/kitsune_p2p/timestamp/src/lib.rs#L38) will properly sort by the i64 including negative values...

Where is it that we are sorting by this display/to_string here?

Reply:
@neonphog in the database. Note that OpOrder is more than just a timestamp, it includes an enum, for a total of 65 bytes of data which can't fit in an i64.

It's just to have a deterministic ordering of ops in the context of dependency resolution. The timestamp is a tie-breaker. If anyone uses negative timestamps, it just means that we'll process those ops out of order, within the class of ops already ordered by the initial enum byte. Which, I don't think is a big deal.

Reply:
aiight : ) seems a little odd, but this pretty much shouldn't ever come up under normal usage, so no worries :+1: